directive_definition_derive/src/lib.rs

```diff
@@ -0,0 +1,125 @@
+use proc_macro::TokenStream;
+
+use quote::quote;
+use syn::{parse, Data, DataStruct, Fields, Type};
+
+#[proc_macro_derive(DirectiveDefinition)]
+pub fn directive_definition_derive(input: TokenStream) -> TokenStream {
+  let ast: syn::DeriveInput = parse(input).unwrap();
+  impl_directive_definition(&ast)
+}
+
+fn get_first_seg_ident_string(path: &syn::TypePath) -> Option<String> {
+    if let Some(seg) = path.path.segments.first() {
+        Some(seg.ident.to_string())
+    } else {
+        None
+    }
+}
+
+fn get_graphql_type(path: &syn::TypePath, is_required: bool) -> String {
+    let ident_string = get_first_seg_ident_string(path);
+    let ident_string_str = ident_string.as_ref().unwrap().as_str();
+    let is_child_required = ident_string_str!= "Option" && ident_string_str != "Vec";
+    let argument_types = match &path.path.segments.first().unwrap().arguments {
+        syn::PathArguments::AngleBracketed(angle_bracketed_args) => {
+            angle_bracketed_args.args.iter().filter_map(|arg|
+                match arg {
+                    syn::GenericArgument::Type(syn::Type::Path(arg_type_path)) => {
+                        Some(get_graphql_type(&arg_type_path, is_child_required))
+                    },
+                    _ => None
+                }
+            ).collect::<Vec<String>>().join(", ")
+        },
+        _ => "".to_string()
+    };
+    let mut graphql_type_str = match ident_string_str {
+        "Option" => format!("{}", argument_types),
+        "BTreeMap" => "[KeyValue]".to_string(),
+        "Vec" => format!("[{}]", argument_types),
+        "Method" => "Method".to_string(),
+        _ => {
+            format!("{}", ident_string.as_ref().unwrap())
+        }
+    };
+    if is_required && ident_string_str != "Option" && ident_string_str != "BTreeMap" {
```

Comment:
Some hardcoded checks here and on line 23  to check that Option, a child of Option, and BTreeMap are not marked required, Vec is marked required, but the Vec argument is not (since the Vec can be empty)

Reply:
Created a draft PR #424 have a look at the implementation. I think structs work properly however, enum support still needs to be added.